File: backend/app/schemas/financial_cushion.py
"""DTO анализа PDF-выписки — зеркала моделей готового пакета financial_cushion."""

from typing import List, Optional

from pydantic import BaseModel, Field, model_validator


class StatementTxDTO(BaseModel):
    """Транзакция выписки, редактируемая пользователем (round-trip JSON)."""

    #: Уникальный ID транзакции (uuid4 hex).
    tx_id: str
    #: Дата операции, YYYY-MM-DD.
    date: str
    #: Категория, определённая алгоритмом.
    group: str
    #: Мерчант / ФИО контрагента.
    counterparty: str
    #: Сумма списания, руб.
    amount: float = Field(gt=0)
    #: Регулярная ли трата по мнению алгоритма.
    is_regular: bool = False
    #: Учитывать ли трату в пересчёте (пользователь может выключить).
    is_active: bool = True
    #: True — регулярный p2p-перевод (кандидат в услуги).
    is_transfer: bool = False
    #: Пользовательская категория услуги для перевода.
    transfer_label: Optional[str] = None
    #: Для переводов: включать ли в подушку.
    include_in_cushion: bool = False


class TransferCandidateDTO(BaseModel):
    """Подсказка UI: регулярный перевод, требующий решения."""

    #: ФИО / счёт получателя перевода.
    counterparty: str
    #: Сумма одного перевода, руб.
    amount: float
    #: Сколько раз повторялся.
    occurrences: int


class StatementAnalysisResponse(BaseModel):
    """Ответ загрузки выписки — пользователь правит его и отправляет назад."""

    #: Полные месяцы периода (например, «июнь 2026»).
    full_months: List[str]
    #: Отброшенные частичные месяцы.
    excluded_months: List[str]
    #: Число полных месяцев (знаменатель формулы).
    months_analyzed: int = Field(ge=1)
    #: Транзакции списаний за полные месяцы.
    transactions: List[StatementTxDTO]
    #: Регулярные переводы-кандидаты в «скрытые подписки».
    transfer_candidates: List[TransferCandidateDTO] = []


class StatementRecalculateRequest(BaseModel):
    """Изменённый пользователем JSON для пересчёта подушки."""

    #: Число полных месяцев (можно не слать, если есть full_months).
    months_analyzed: Optional[int] = Field(default=None, ge=1)
    #: Подписи полных месяцев.
    full_months: List[str] = []
    #: Отброшенные частичные месяцы.
    excluded_months: List[str] = []
    #: Транзакции с флагами пользователя.
    transactions: List[StatementTxDTO]

    @model_validator(mode="after")
    def _check_denominator(self) -> "StatementRecalculateRequest":
        if self.months_analyzed is None and not self.full_months:
            raise ValueError(
                "Нужен months_analyzed либо непустой full_months "
                "(знаменатель формулы пересчёта)"
            )
        return self


class StatementGroupDTO(BaseModel):
    """Итог по одной группе расходов."""

    #: Название категории.
    name: str
    #: Среднее в месяц, руб.
    monthly_amount: float
    #: Число операций в месяц (в среднем).
    ops_count: int
    #: Всего потрачено за все полные месяцы, руб.
    total_amount: float = 0.0
    #: Пояснение, что входит в группу.
    description: str = ""
    #: Список операций «контрагент — сумма» (по убыванию суммы).
    operations: List[str] = []


class StatementRecalculateResponse(BaseModel):
    """Результат пересчёта подушки по правкам пользователя."""

    #: Число полных месяцев (знаменатель).
    months_analyzed: int
    #: Итоговая сумма подушки: [активные траты] / [полные месяцы].
    cushion_total: float
    #: Категории, включённые в подушку (по убыванию суммы).
    groups: List[StatementGroupDTO]
    #: Категории вне подушки (исключены пользователем / нерегулярные).
    excluded_groups: List[StatementGroupDTO]


class FamilyCushionRatingItem(BaseModel):
    """Строка рейтинга семьи: финподушка одного участника."""

    #: Идентификатор пользователя.
    user_id: int
    #: Имя участника.
    user_name: str
    #: Итог его финподушки (0, если выписка не загружалась).
    cushion_total: float
    #: Число полных месяцев в его расчёте.
    months_analyzed: Optional[int] = None


__all__ = [
    "StatementTxDTO",
    "TransferCandidateDTO",
    "StatementAnalysisResponse",
    "StatementRecalculateRequest",
    "StatementGroupDTO",
    "StatementRecalculateResponse",
    "FamilyCushionRatingItem",
]

File: frontend/src/shared/api/financialCushion.ts
import { API_URL, getStoredToken, request } from './base'
import type { APIResponse } from './types'

export interface StatementTx {
  tx_id: string
  date: string
  group: string
  counterparty: string
  amount: number
  is_regular: boolean
  is_active: boolean
  is_transfer: boolean
  transfer_label: string | null
  include_in_cushion: boolean
}

export interface TransferCandidate {
  counterparty: string
  amount: number
  occurrences: number
}

export interface StatementAnalysis {
  full_months: string[]
  excluded_months: string[]
  months_analyzed: number
  transactions: StatementTx[]
  transfer_candidates: TransferCandidate[]
}

export interface StatementGroup {
  name: string
  monthly_amount: number
  ops_count: number
  total_amount: number
  description: string
  operations: string[]
}

export interface StatementRecalcRequest {
  months_analyzed?: number | null
  full_months: string[]
  excluded_months: string[]
  transactions: StatementTx[]
}

export interface StatementRecalcResult {
  months_analyzed: number
  cushion_total: number
  groups: StatementGroup[]
  excluded_groups: StatementGroup[]
}

export interface FamilyCushionRatingItem {
  user_id: number
  user_name: string
  cushion_total: number
  months_analyzed: number | null
}

export async function uploadStatement(
  file: File,
): Promise<APIResponse<StatementAnalysis>> {
  // Отдельный fetch для multipart: Content-Type с boundary выставляет
  // браузер, поэтому общий request() (жёстко ставящий application/json)
  // для FormData не подходит.
  try {
    const headers: Record<string, string> = {}
    const token = getStoredToken()
    if (token) {
      headers.Authorization = `Bearer ${token}`
    }

    const response = await fetch(
      `${API_URL}/financial-cushion/upload-statement`,
      {
        method: 'POST',
        headers,
        body: buildStatementFormData(file),
      },
    )

    const rawPayload = await response.json()
    const payload = rawPayload as APIResponse<StatementAnalysis> & {
      detail?: Array<{ msg?: string }>
    }

    if (!response.ok) {
      const validationMessage = payload.detail
        ?.map((item) => item.msg)
        .filter(Boolean)
        .join(' ')

      return {
        error: true,
        message:
          payload.message ?? validationMessage ?? 'Ошибка запроса к серверу.',
        data: null,
        type: payload.type ?? `http_${response.status}`,
      }
    }

    return payload
  } catch {
    return {
      error: true,
      message: 'Не удалось подключиться к серверу.',
      data: null,
      type: 'network_error',
    }
  }
}

export function recalculateStatement(
  payload: StatementRecalcRequest,
): Promise<APIResponse<StatementRecalcResult>> {
  return request<StatementRecalcResult>('/financial-cushion/recalculate', {
    method: 'POST',
    body: JSON.stringify(payload),
  })
}

export function getLastAnalysis(): Promise<APIResponse<StatementAnalysis | null>> {
  return request<StatementAnalysis | null>('/financial-cushion/last', {
    method: 'GET',
  })
}

export function getFamilySummary(): Promise<APIResponse<FamilyCushionRatingItem[]>> {
  return request<FamilyCushionRatingItem[]>('/financial-cushion/family-summary', {
    method: 'GET',
  })
}

export function resetLastAnalysis(): Promise<APIResponse<null>> {
  return request<null>('/financial-cushion/last', {
    method: 'DELETE',
  })
}

function buildStatementFormData(file: File): FormData {
  const formData = new FormData()
  formData.append('file', file, file.name || 'statement.pdf')
  return formData
}
